File: jvm/src/main/java/edu/kit/kastel/vads/compiler/semantic/SemanticAnalysis.java
package edu.kit.kastel.vads.compiler.semantic;

import de.mr_pine.c0ne.semantic.MainMethodAnalysis;
import edu.kit.kastel.vads.compiler.parser.ast.ProgramTree;
import edu.kit.kastel.vads.compiler.parser.visitor.RecursivePostorderVisitor;
import kotlin.Unit;

public class SemanticAnalysis {

    private final ProgramTree program;

    public SemanticAnalysis(ProgramTree program) {
        this.program = program;
    }

    public void analyze() {
        this.program.accept(new RecursivePostorderVisitor<>(new IntegerLiteralRangeAnalysis()), new Namespace<>());
        this.program.accept(new RecursivePostorderVisitor<>(new VariableStatusAnalysis()), new Namespace<>());
        this.program.accept(new RecursivePostorderVisitor<>(new ReturnAnalysis()), new ReturnAnalysis.ReturnState());
        this.program.accept(new RecursivePostorderVisitor<>(new MainMethodAnalysis()), Unit.INSTANCE);
    }

}

File: jvm/src/main/java/de/mr_pine/c0ne/semantic/SemanticAnalysis.kt
package de.mr_pine.c0ne.semantic

import de.mr_pine.c0ne.parser.ast.ProgramTree
import de.mr_pine.c0ne.parser.visitor.RecursivePostorderVisitor
import de.mr_pine.c0ne.semantic.ReturnAnalysis.ReturnState

class SemanticAnalysis(private val program: ProgramTree) {
    fun analyze() {
        this.program.accept(
            RecursivePostorderVisitor(
                IntegerLiteralRangeAnalysis()
            ), Unit
        )
        this.program.accept(
            VariableStatusAnalysis(), VariableStatusAnalysis.VariableStatus.initial
        )
        this.program.accept(
            RecursivePostorderVisitor(TypeCheckAnalysis()), mutableListOf()
        )
        this.program.accept(RecursivePostorderVisitor(BreakContinueAnalysis()), mutableListOf())
        this.program.accept(
            RecursivePostorderVisitor(ReturnAnalysis()), ReturnState()
        )
        this.program.accept(
            RecursivePostorderVisitor(MainMethodAnalysis()), Unit
        )
    }
}
